Cabal/tests/PackageTests/Tests.hs

```diff
@@ -0,0 +1,157 @@
+module PackageTests.Tests(tests) where
+
+import PackageTests.PackageTester
+
+import qualified PackageTests.BenchmarkStanza.Check
+import qualified PackageTests.TestStanza.Check
+import qualified PackageTests.DeterministicAr.Check
+import qualified PackageTests.TestSuiteTests.ExeV10.Check
+
+import Distribution.Simple.Utils (withFileContents)
+import Control.Monad
+
+import Data.Version
+import Test.Tasty (TestTree, testGroup, mkTimeout, localOption)
+import Test.Tasty.HUnit (testCase)
+
+-- TODO: turn this into a "test-defining writer monad".
+-- This will let us handle scoping gracefully.
+tests :: SuiteConfig -> [TestTree]
+tests config =
+  tail [ undefined
+
+  -- External tests
+  , tc "BenchmarkStanza"  PackageTests.BenchmarkStanza.Check.suite
+  , tc "TestStanza"       PackageTests.TestStanza.Check.suite
+  , tc "DeterministicAr"  PackageTests.DeterministicAr.Check.suite
+
+  -- Test suite tests
+  , testGroup "TestSuiteTests"
+    [ testGroup "ExeV10"
+      (PackageTests.TestSuiteTests.ExeV10.Check.tests config)
+    , testGroup "LibV09" $
+      let
+        tcs :: FilePath -> TestM a -> TestTree
+        tcs name m
+            = testCase name (runTestM config ("TestSuiteTests/LibV09")
+                                             (Just name) m)
+      in [ tcs "Build" $ cabal_build ["--enable-tests"]
+         , localOption (mkTimeout $ 10 ^ (8 :: Int))
+         . tcs "Deadlock" $ do
+            cabal_build ["--enable-tests"]
+            shouldFail $ cabal "test" []
+         ]
+    ]
+
+  -- Inline tests
+  , tc "BenchmarkExeV10" $ cabal_build ["--enable-benchmarks"]
+
+  , tc "BenchmarkOptions" $ do
```

Comment:
Sure, and I definitely agree that it should be hard to "do the wrong thing" (e.g. forget to test exit status). There are certain things that should always be "implicitly-check-this-unless-told-otherwise" and running ./Setup would definitely be one of those, I think.

It can definitely wait since it requires quite a lot of thought, but it'd be nice to have some sort of comment for each test, just briefly explaining what it's supposed to be testing. (I guess this may not actually be knowable in all cases, though...)